### PR TITLE
Added ADR service discovery

### PR DESCRIPTION
**Pull request by Isilud**

### CODE REVIEW

**Review by ika108 — APPROVED** [review]

File: docs/ADR/013-transformation-services-infra.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+#### Benefits
+- Act only when needed.
+- Detection of a service absence/crash is left to Consul.
+- We don't need to reconnect to the service with each request.
+
+#### Drawbacks
+- Not sure it's Consul task.
+- Need a cache to keep the list in the API.
+- In case of multiple API, Consul may update only one per request.
```
Comment (RaboliotLeGris) [inline docs/ADR/013-transformation-services-infra.md:30]: Not sure of what you mean by that ?

File: docs/ADR/013-transformation-services-infra.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+![Diagram Option 1](../diagram/transformers_discovery/ADR_13_opt1.jpg)
+
+#### Benefits
+- Act only when needed.
+- Detection of a service absence/crash is left to Consul.
+- We don't need to reconnect to the service with each request.
+
+#### Drawbacks
+- Not sure it's Consul task.
+- Need a cache to keep the list in the API.
```
Comment (RaboliotLeGris) [inline docs/ADR/013-transformation-services-infra.md:29]: Why N

File: docs/ADR/013-transformation-services-infra.md (hunk 1 above)
Comment (Isilud) [reply docs/ADR/013-transformation-services-infra.md:30]: If we depend of a "POST" request made by Consul on our API for updating our list of services, then we will need a place to stock those. It would be a "cache" for the API to keep track of existing service and their addresses.

File: docs/ADR/013-transformation-services-infra.md (hunk 2 above)
Comment (Isilud) [reply docs/ADR/013-transformation-services-infra.md:29]: I don't like the idea of Consul being in charge of giving the services list by itself. I would prefer for our services to do a request when needed to get it.

Comment (RaboliotLeGris) [reply docs/ADR/013-transformation-services-infra.md:29]: Then it makes you API less stateless no ?

Comment (ika108) [reply docs/ADR/013-transformation-services-infra.md:29]: So... Statemore ? 🤭

Comment (Isilud) [reply docs/ADR/013-transformation-services-infra.md:29]: It didn't occur to me before but yes, it does.

File: docs/ADR/013-transformation-services-infra.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+### 2. API ask Consul
+
+The API is managing the user requests. We could fetch the list of existing transformation/services for each transformation request, because Squarescale register new service automatically on Consul. 
+![Diagram Option 2](../diagram/transformers_discovery/ADR_13_opt2.jpg)
+
+#### Benefits
+- Ensure informations are up to date when needed.
+- Detection of a service absence/crash is left to Consul.
+- We don't need to cache the adress/client.
+- Simple implementation.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:42]: - We can easily replace consul by another solution in  he future, if needed.

File: docs/ADR/013-transformation-services-infra.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#### Benefits
+- Act only when needed.
+- Detection of a service absence/crash is left to Consul.
+- We don't need to reconnect to the service with each request.
+
+#### Drawbacks
+- Not sure it's Consul task.
+- Need a cache to keep the list in the API.
+- In case of multiple API, Consul may update only one per request.
+- Harder implementation.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:31]: + We need to extend consul for that. do we want to take part of Consul devlopment and buy all problems this wonderful projest already has ?

File: docs/ADR/013-transformation-services-infra.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+![Diagram Option 2](../diagram/transformers_discovery/ADR_13_opt2.jpg)
+
+#### Benefits
+- Ensure informations are up to date when needed.
+- Detection of a service absence/crash is left to Consul.
+- We don't need to cache the adress/client.
+- Simple implementation.
+
+#### Drawbacks
+- APIs will make lots of call.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:45]: I'm not sure of that. We only need to make a request to consul in two cases :  

1. The first time we try to reach that service
2. The connection tineouts (service has probably moved away)

In all other cases, we won't make call to consul (we juste need to keep the IP address in memory, so we can reconnect later)

File: docs/ADR/013-transformation-services-infra.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- We will need to recreate a RPC client for each request.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:46]: Absolutely not. This point is not related to service discovery at all.

File: docs/ADR/013-transformation-services-infra.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+#### Points of attention
+We can also maintain a cache to keep the list in the API. Then, we can reduce API calls to Consul and service reconnection (may increase time to detect fault).
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:49]: OK, we can also detail a but more, saying : 

> We keep IP addresses of each service we discover thank to consul within serice memory, until it soed not work anymore

File: docs/ADR/013-transformation-services-infra.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+### 3. Services ask API
+
+Another way to update the list of services is by letting them annonce themselves on an API endpoint when they are up and running. To do so, the API need maintain a list of running services.
+![Diagram Option 3](../diagram/transformers_discovery/ADR_13_opt3.jpg)
+
+#### Benefits
+- Remove calls between API and Consul.
+- We don't need to recreate a RPC client for each request.
+- Simple implementation.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:59]: Pretty sure this is not simple ! You'll need to manage clustering and redundancy. Good luck with that !

File: docs/ADR/013-transformation-services-infra.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+#### Drawbacks
+- No warning when a service crash.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:62]: This is clearly not the role of a service discovery

File: docs/ADR/013-transformation-services-infra.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+- In case of multiple API, the service will only contact one.
```
Comment (vlamy) [inline docs/ADR/013-transformation-services-infra.md:63]: This is clearly a No go here. Doesn'it ?

**Review by vlamy — CHANGES_REQUESTED** [review]

File: docs/ADR/013-services-discovery.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+Sans le SDK consul.
+```go
+    address := "gray-server-transformer.service.consul" // Can be retrieve by env var
+    conn, err := grpc.Dial(address, opts)
+    if err != nil {
+        log.Error("Cannot open TCP connection with transformer server :", err)
+    }
+```
+
+### 2. Directly ask Consul for services addresses, using SDK
```
Comment (RaboliotLeGris) [inline docs/ADR/013-services-discovery.md:37]: Oh well, this one seems nice

**Review by vlamy — APPROVED** [review]
